presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java

```diff
@@ -321,7 +334,102 @@ private static int hashBytes(int initialValue, Slice bytes)
         return Optional.of(new HiveBucketFilter(builder.build()));
     }
 
-    private static OptionalInt getHiveBucket(Table table, Map<ColumnHandle, NullableValue> bindings)
+    private static Optional<Set<Integer>> getHiveBuckets(Table table, Map<ColumnHandle, Set<NullableValue>> bindings)
+    {
+        ImmutableSet.Builder<Integer> builder = ImmutableSet.builder();
+        Optional<List<Pair<String, TypeInfo>>> typeInfos = getBucketColumnTypeInfos(table);
+
+        if (!typeInfos.isPresent()) {
+            return Optional.empty();
+        }
+
+        if (!getHiveBucketsHelper(bindings, new LinkedHashMap<>(), bindings.keySet().stream().collect(Collectors.toList()), 0,
+                    typeInfos.get(), table.getStorage().getBucketProperty().get().getBucketCount(), builder)) {
+            return Optional.empty();
+        };
+
+        return Optional.of(builder.build());
+    }
+
+    private static boolean getHiveBucketsHelper(
+            Map<ColumnHandle, Set<NullableValue>> bindings,
+            Map<String, NullableValue> currentBinding,
+            List<ColumnHandle> columnHandles,
+            int index,
+            List<Pair<String, TypeInfo>> bucketColTypeInfo,
+            int bucketCount,
+            ImmutableSet.Builder<Integer> builder)
+    {
+        if (index == columnHandles.size()) {
+            OptionalInt singleBucket = getHiveBucket(bucketCount, bucketColTypeInfo, currentBinding);
+            if (singleBucket.isPresent()) {
+                builder.add(singleBucket.getAsInt());
+                return true;
+            } else {
+                return false;
+            }
+        }
+        else {
+            ColumnHandle handle = columnHandles.get(index);
+            for (NullableValue val : bindings.get(handle)) {
+                currentBinding.put(((HiveColumnHandle)handle).getName(), val);
+                if (!getHiveBucketsHelper(bindings, currentBinding, columnHandles, index + 1, bucketColTypeInfo, bucketCount, builder)) {
+                    return false;
+                };
+            }
+            return true;
+        }
+    }
+
+    private static OptionalInt getHiveBucket(int bucketCount, List<Pair<String, TypeInfo>> bucketColumnTypeInfos, Map<String, NullableValue> bindings)
+    {
+        if (bindings.isEmpty()) {
+            return OptionalInt.empty();
+        }
+
+        // Check that we have bindings for all bucket columns
+        if (bindings.size() != bucketColumnTypeInfos.size()) {
+            return OptionalInt.empty();
+        }
+
+        ImmutableList.Builder<TypeInfo> typeInfos = ImmutableList.builder();
```

Comment:
The list of types is the same for each set of values. No need to re-compute it again and again. Consider refactoring the logic to (1) get a list of types; (2) check if there are bindings for all necessary columns; (3) generate all possible permutations of bucketed_by column values and compute buckets for each.

```
    private static Optional<Set<Integer>> getHiveBuckets(Table table, Map<ColumnHandle, Set<NullableValue>> bindings)
    {
        Optional<List<Pair<String, TypeInfo>>> nameToType = getBucketColumnTypeInfos(table);

        if (!nameToType.isPresent()) {
            return Optional.empty();
        }

        Map<String, Set<NullableValue>> nameToBindings = bindings.entrySet().stream()
                .collect(toImmutableMap(entry -> ((HiveColumnHandle) entry.getKey()).getName(), Entry::getValue));

        ImmutableList.Builder<Set<NullableValue>> orderedBindingsBuilder = ImmutableList.builder();
        for (Pair<String, TypeInfo> typeInfoPair : nameToType.get()) {
            String columnName = typeInfoPair.getKey();
            if (!nameToBindings.containsKey(columnName)) {
                return Optional.empty();
            }
            orderedBindingsBuilder.add(nameToBindings.get(columnName));
        }

        List<Set<NullableValue>> orderedBindings = orderedBindingsBuilder.build();

        int bucketCount = table.getStorage().getBucketProperty().get().getBucketCount();
        List<TypeInfo> types = nameToType.get().stream()
                .map(Pair::getRight)
                .collect(toImmutableList());

        ImmutableSet.Builder<Integer> buckets = ImmutableSet.builder();
        getHiveBuckets(new Object[types.size()], 0, orderedBindings, bucketCount, types, buckets);

        return Optional.of(buckets.build());
    }

    private static void getHiveBuckets(
            Object[] values,
            int valuesCount,
            List<Set<NullableValue>> bindings,
            int bucketCount,
            List<TypeInfo> typeInfos,
            ImmutableSet.Builder<Integer> buckets)
    {
        if (valuesCount == typeInfos.size()) {
            buckets.add(getHiveBucket(bucketCount, typeInfos, values));
            return;
        }

        for (NullableValue value : bindings.get(valuesCount)) {
            values[valuesCount] = value.getValue();
            getHiveBuckets(values, valuesCount + 1, bindings, bucketCount, typeInfos, buckets);
        }
    }
```

Reply:
Thanks! Done.